libraries/botbuilder-lg/src/templates.ts

```diff
@@ -312,24 +342,77 @@ export class Templates implements Iterable<Template> {
     * @returns Return the new lg file.
     */
     public deleteTemplate(templateName: string): Templates {
-        const template: Template = this.items.find((u: Template): boolean => u.name === templateName);
-        if (template === undefined) {
-            return this;
+        const templateIndex = this.items.findIndex((u: Template): boolean => u.name === templateName);
+        if (templateIndex >= 0) {
+            const template = this.items[templateIndex];
+            this.clearDiagnostic();
+
+            const startLine = template.sourceRange.range.start.line - 1;
+            const stopLine = template.sourceRange.range.end.line - 1;
+            this.content = this.replaceRangeContent(this.content, startLine, stopLine, undefined);
+            this.adjustRangeForDeleteTemplate(template);
+            this.items.splice(templateIndex, 1);
+            new StaticChecker(this).check().forEach((u): number => this.diagnostics.push(u));
         }
 
-        let startLine = template.sourceRange.range.start.line - 1;
-        let stopLine = template.sourceRange.range.end.line - 1;
-
-        const newContent: string = this.replaceRangeContent(this.content, startLine, stopLine, undefined);
-        this.initialize(TemplatesParser.parseText(newContent, this.id, this.importResolver));
-
         return this;
     }
 
     public toString(): string {
         return this.content;
     }
 
+    private appendDiagnosticWithOffset(diagnostics: Diagnostic[], offset: number): void {
+        if (diagnostics) {
+            diagnostics.forEach((u): void => {
+                u.range.start.line += offset;
+                u.range.end.line += offset;
+                this.diagnostics.push(u);
+            });
+        }
+    }
+
+    private adjustRangeForUpdateTemplate(oldTemplate: Template, newTemplate: Template): void {
+        const lineOffset = newTemplate.sourceRange.range.end.line - newTemplate.sourceRange.range.start.line
+        - (oldTemplate.sourceRange.range.end.line - oldTemplate.sourceRange.range.start.line);
```

Comment:
This would be clearer if you did a const newRange = newTemplate.sourceRange.range and const oldRange = oldTemplate.soureceRange first and rewrote that in terms of this.

Reply:
#Solved